### ISSUE TITLE
Training on custom dataset, Inference nothing

### ISSUE THREAD

**Issue by github2016-yuan** [google/automl#521]
I have trained voc 2012 and coco val2017. The pb files I get are well and can detect bboxes in images when inference. That is to say I konw how to use efficientdet-d0 to train on public datasets.
Now I want to train a detector of bus so I extract the bus class (yes just one class) from coco train2017. Then I get bus.json and 3952 bus images. Next I do what I have done when training on coco val2017.
After 90 epoch, the final loss is 0.56 and I also use tensorboard to visualize the losses.
**box_loss: 1.88e-3
cls_loss: 2.70e-3
det_loss: 0.0051
loss: 0.56**
It seems ok and I convert ckpt into pb file and then use pb file for inference. No error occurs but it can not detect anything. I run into the question before  and solve it by adding more epochs. Here I have already trained for 90 epochs and should have some results when inference.
Here are my cms:

**1.training:**
--mode=train
--training_file_pattern=cocotfrecordbus/train/* 
--model_name=efficientdet-d0
--model_dir=cocobusmodels
--backbone_ckpt=efficientnet-b0  
--train_batch_size=8 
--num_examples_per_epoch=3952 
--num_epochs=90 
--hparams="num_classes=1,moving_average_decay=0,precision=mixed_float16"

**2. converting ckpt to pb**
--runmode=saved_model 
--model_name=efficientdet-d0  
--ckpt_path=cocobusmodels
--saved_model_dir=savedcocomodeldir   
--hparams="num_classes=1,moving_average_decay=0,precision=mixed_float16"

**3.inference:**
--runmode=saved_model_infer  
--model_name=efficientdet-d0  
--saved_model_dir=savedcocomodeldir  
--hparams="num_classes=1,moving_average_decay=0,precision=mixed_float16" 
--input_image=bus4.jpg   
--output_image_dir=./tmp/

Similar issue: #499 
Need your help please.

**Comment by sandstorm12** [comment]
I have the same issue. I have trained the checkpoint exactly like commands on fine-tuning on Pascal. All codes work but inference draws no boxes on test images.

tensorflow==2.2.0
Cuda compilation tools, release 10.1, V10.1.105

**Comment by MustafaUysal160207077** [comment]
I tried to train on Google Colab. I have the same issue.

**Comment by github2016-yuan** [comment]
@mingxingtan something wrong with the code or our training on custom data?

**Comment by mingxingtan** [comment]
Can someone provide me some example data, (or a pretrained model and some test images)?  I would like to debug it to see what's wrong. Thanks!

**Comment by github2016-yuan** [comment]
@mingxingtan 
[BaiduCloud](https://pan.baidu.com/s/1W_jy4V0erK5TWvfX-EJKoA )
password:  **o12s**

Here is my small dataset of bus. I create it from coco train2017. Maybe it is helpful to debug.

**Comment by sandstorm12** [comment]
I have trained for a little longer and the problem is solved. I had fine-tuned the model for only 1 epoch first time and had this problem. Training for 10 epochs solved my issue. Maybe reduce the confidence threshold, see what are your outputs?

**Comment by adarsh0906** [comment]
> 
> 
> I have trained for a little longer and the problem is solved. I had fine-tuned the model for only 1 epoch first time and had this problem. Training for 10 epochs solved my issue. Maybe reduce the confidence threshold, see what are your outputs?

i couldn't get confidence at all the prediction outputs an empty array

**Comment by adarsh0906** [comment]
https://drive.google.com/drive/folders/1--GUV5UdefeUxMukn4oI5U_TNKp6z7xb

here is the link for the model trained on cutomdataset on efficientdet d1 model the image for inference is 
![hw250](https://user-images.githubusercontent.com/18246781/84928756-8a764a80-b0ec-11ea-92f2-70b0e8395962.jpg)
![M3025100](https://user-images.githubusercontent.com/18246781/84928759-8ba77780-b0ec-11ea-93c6-7e9a9e2c8946.jpg)




> 
> 
> Can someone provide me some example data, (or a pretrained model and some test images)? I would like to debug it to see what's wrong. Thanks!

**Comment by github2016-yuan** [comment]
@mingxingtan Have you solved the problem about invalid pb file when inference?

**Comment by mingxingtan** [comment]
@github2016-yuan 

I am not able to download your data, but after reading through the code again and some other issues #391, I believe you should set num_classes=2. Could you give it a try?

I will try to generate my pascal data with only one class to test that.

**Comment by kerasus41** [comment]
> @github2016-yuan
> 
> I am not able to download your data, but after reading through the code again and some other issues #391, I believe you should set num_classes=2. Could you give it a try?
> 
> I will try to generate my pascal data with only one class to test that.

can u please look at to reply of @adarsh0906 ? he had set t num_classes=4 but issue is not fixed when I try inference.

**Comment by adarsh0906** [comment]
> 
> 
> > @github2016-yuan
> > I am not able to download your data, but after reading through the code again and some other issues #391, I believe you should set num_classes=2. Could you give it a try?
> > I will try to generate my pascal data with only one class to test that.
> 
> can u please look at to reply of @adarsh0906 ? he had set t num_classes=4 but issue is not fixed when I try inference.

i had four class(car,bus,truck,motorcycle) during training,should i make num_classess = 5 during training for num_class+background?

**Comment by mingxingtan** [comment]
That's right @adamsyu

**Comment by adarsh0906** [comment]
> 
> 
> That's right @adamsyu

it is working while training number of classes should be number of class +1 to make detection when inference

**Comment by github2016-yuan** [comment]
It seems that num_classes + 1 could solve this problem. I download the newest code and set num_classes + 1 when training on my own dataset. After 30 epochs, pb file that I get is still valid.
Could someone share cmds anout training and inference?
@adarsh0906

**Comment by github2016-yuan** [comment]
It seems that num_classes + 1 could solve this problem. I download the newest code and set num_classes + 1 when training on my own dataset. After 30 epochs, pb file that I get is still valid.
Could someone share cmds anout training and inference?
@adarsh0906

**Comment by adarsh0906** [comment]
> 
> 
> It seems that num_classes + 1 could solve this problem. I download the newest code and set num_classes + 1 when training on my own dataset. After 30 epochs, pb file that I get is still valid.
> Could someone share cmds anout training and inference?
> @adarsh0906

The cmds are same just the num_classes in hparams is num_classes in your data + 1 for both training and inference.

num_class = num_classes +1
`
!python main.py --mode=train_and_eval \
    --training_file_pattern=tfrecord/{file_pattern} \
    --validation_file_pattern=tfrecord/{file_pattern} \
    --model_name={MODEL} \
    --model_dir=/tmp/model_dir/{MODEL}-scratch  \
    --backbone_ckpt={backbone_name} \
    --train_batch_size=4 \
    --eval_batch_size=4 --eval_samples={images_per_epoch}  \
    --num_examples_per_epoch={images_per_epoch}  --num_epochs=1  \
    --hparams="num_classes=num_class,moving_average_decay=0,mixed_precision=true"
`

**Comment by adarsh0906** [comment]
@mingxingtan  there is another issue while training in colab when the mode is train and eval the ram is increasing is that strange or its normal thing according to the code

**Comment by mingxingtan** [comment]
@adarsh0906 
It might be a bug. I don't expect memory increasing.  
Is it specific to "train_and_eval"? I usually started two jobs: one for train and the other for eavl, and did not observe this issue.

**Comment by adarsh0906** [comment]
> 
> 
> @adarsh0906
> It might be a bug. I don't expect memory increasing.
> Is it specific to "train_and_eval"? I usually started two jobs: one for train and the other for eavl, and did not observe this issue.

it is specific to train and eval in colab ?

**Comment by mingxingtan** [comment]
I suspect the memory increase is caused  by the switch between train and eval .

How about you change [this line ](https://github.com/google/automl/blob/21aa73cfe0d20a7c671c88274769c33be59edf92/efficientdet/main.py#L373)to something like:  

`for cycle in range(5, config.num_epochs, 5)`

so it evals every 5 epochs. You can also set "eval_after_training" flag, which would only do eval after all the training.

**Comment by xiaoxin05** [comment]
> tensorflow==2.2.0

hello, Are you under Ubuntu environment

**Comment by adarsh0906** [comment]
> 
> 
> I suspect the memory increase is caused by the switch between train and eval .
> 
> How about you change [this line ](https://github.com/google/automl/blob/21aa73cfe0d20a7c671c88274769c33be59edf92/efficientdet/main.py#L373)to something like:
> 
> `for cycle in range(5, config.num_epochs, 5)`
> 
> so it evals every 5 epochs. You can also set "eval_after_training" flag, which would only do eval after all the training.

i tried using eval after training it worked thank you  

@mingxingtan #548 can you help with this also
